compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -0,0 +1,251 @@
+import DA.Daml.Compiler.Dar
+import DA.Daml.Compiler.DecodeDar (DecodedDalf(..), decodeDalf)
+import DA.Daml.Compiler.ExtractDar (ExtractedDar(..), extractDar)
+import qualified DA.Daml.LF.Ast as LF
+import DA.Daml.Options.Types
+import DA.Daml.Package.Config
+import qualified DA.Pretty
+import Data.Aeson (eitherDecodeFileStrict', encode)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import Data.List.Extra
+import qualified Data.Set as Set
+import qualified Data.Text.Extended as T
+import Development.IDE.Types.Location
+import GHC.Fingerprint
+import System.Directory.Extra
+import System.FilePath
+import System.IO.Extra
+
+-- Dependency Database Layout
+-----------------------------
+-- Here is an exemplary dependecy database:
+-- .
+-- ├── data-deps
+-- │   ├── configs
+-- │   │   └── proj2-0.0.1.conf
+-- │   ├── dalfs
+-- │   │   ├── 057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba
+-- │   │   │   └── daml-stdlib-DA-Internal-Down-057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba.dalf
+-- │   │   ├── 3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
+-- │   │   │   └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3.dalf
+-- │   │   └── e9eeb8cf890e5a15bc65031f1a8373a4bb7f89229473dabaa595c25d35768b21
+-- │   │       └── daml-prim-e9eeb8cf890e5a15bc65031f1a8373a4bb7f89229473dabaa595c25d35768b21.dalf
+-- │   ├── mains
+-- │   │   └── 3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
+-- │   │       └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3.dalf
+-- │   ├── sdk-version
+-- │   └── sources
+-- │       └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
+-- │           ├── Baz.daml
+-- │           ├── Baz.hi
+-- │           └── Baz.hie
+-- ├── deps
+-- │   ├── configs
+-- │   │   └── daml-script-0.0.0.conf
+-- │   ├── dalfs
+-- │   │   ├── 057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba
+-- │   │   │   └── daml-stdlib-DA-Internal-Down-057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba.dalf
+-- │   │   ├── 40f452260bef3f29dede136108fc08a88d5a5250310281067087da6f0baddff7
+-- │   │   │   └── daml-prim-DA-Types-40f452260bef3f29dede136108fc08a88d5a5250310281067087da6f0baddff7.dalf
+-- │   ├── mains
+-- │   │   └── bf4a8f6f897b4d9639bab0f384ea7c06a7ad4b6b856cb0b511e1d79c9c46ecf2
+-- │   │       └── daml-script-0.0.0-bf4a8f6f897b4d9639bab0f384ea7c06a7ad4b6b856cb0b511e1d79c9c46ecf2.dalf
+-- │   ├── sdk-version
+-- │   └── sources
+-- │       └── daml-script-0.0.0-bf4a8f6f897b4d9639bab0f384ea7c06a7ad4b6b856cb0b511e1d79c9c46ecf2
+-- │           └── Daml
+-- │               ├── Script
+-- │               │   ├── Free.daml
+-- │               │   ├── Free.hi
+-- │               │   └── Free.hie
+-- │               ├── Script.daml
+-- │               ├── Script.hi
+-- │               └── Script.hie
+-- └── fingerprint.json
```

Comment:
What happens if a dalf is pulled in as a `dependency` from one DAR and a `data-dependency` from another DAR? Maybe a good way to avoid that ambiguity is to not have separate directories and instead use a marker file?

Reply:
there's a marker now.